data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,193 @@
+          .entrySet())
+      {
+        PathSpec pathSpec = e.getKey();
+        Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>> olderCheckContextAndResolvedProperty =
+            olderResolvedPropertiesMap.get(pathSpec);
+        Map<String, Object> olderResolvedProperties = olderCheckContextAndResolvedProperty.getValue();
+
+        if (newerResolvedPropertiesMap.containsKey(pathSpec))
+        {
+          // If older schema node and newer schema node have the same pathSpec,
+          // they may or may not contain the same annotation namespace as SchemaAnnotationHandler, we need to further check.
+          Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>
+              newerCheckContextAndResolvedProperty = newerResolvedPropertiesMap.get(pathSpec);
+          Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+
+          // If olderResolvedProperties or newerResolvedProperties contains the same namespace as the provided SchemaAnnotationHandler,
+          // we do the annotation check
+          if (olderResolvedProperties.containsKey(annotationNamespace) || newerResolvedProperties.containsKey(
+              annotationNamespace))
+          {
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, newerResolvedProperties,
+                    olderCheckContextAndResolvedProperty.getKey(), newerCheckContextAndResolvedProperty.getKey());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
+            results.add(result);
+          }
+          newerResolvedPropertiesMap.remove(pathSpec);
+        }
+        else
+        {
+          if (olderResolvedProperties.containsKey(annotationNamespace))
+          {
+            // olderResolvedPropertiesMap has a PathSpec which the newResolvedPropertiesMap does not have,
+            // it could be either newer schema removed annotation or the field is not existed, pass
+            // an empty newerResolvedPropertiesMap and empty newerAnnotationContext to the annotation check.
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, new HashMap<>(),
+                    olderCheckContextAndResolvedProperty.getKey(), new SchemaAnnotationHandler.AnnotationCheckContext());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
```

Comment:
this duplication can be avoided if we added a method on the handler to check.

Reply:
add a method on the handler.